src/message.rs

```diff
@@ -0,0 +1,129 @@
+    ///
+    /// # Safety
+    ///
+    /// There is no guarantee the provided bytes will be the product of a cryptographically secure
+    /// hash. Using insecure messages might compromise the security of the signature.
+    pub unsafe fn from_bytes_unchecked(bytes: [u8; Self::LEN]) -> Self {
+        Self(bytes.into())
+    }
+
+    /// Add a conversion from arbitrary slices into owned
+    ///
+    /// # Safety
+    ///
+    /// This function will not panic if the length of the slice is smaller than
+    /// `Self::LEN`. Instead, it will cause undefined behavior and read random
+    /// disowned bytes.
+    ///
+    /// This function extends the unsafety of [`Self::from_bytes_unchecked`].
+    pub unsafe fn from_slice_unchecked(bytes: &[u8]) -> Self {
+        Self(Bytes32::from_slice_unchecked(bytes))
+    }
+
+    /// Copy-free reference cast
+    ///
+    /// # Safety
+    ///
+    /// Inputs smaller than `Self::LEN` will cause undefined behavior.
+    ///
+    /// This function extends the unsafety of [`Self::from_bytes_unchecked`].
+    pub unsafe fn as_ref_unchecked(bytes: &[u8]) -> &Self {
+        // The interpreter will frequently make references to keys and values using
+        // logically checked slices.
+        //
+        // This function will avoid unnecessary copy to owned slices for the interpreter
+        // access
+        &*(bytes.as_ptr() as *const Self)
+    }
+}
+
+impl Deref for Message {
+    type Target = [u8; Message::LEN];
+
+    fn deref(&self) -> &[u8; Message::LEN] {
+        self.0.deref()
+    }
+}
+
+impl AsRef<[u8]> for Message {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl From<Message> for [u8; Message::LEN] {
+    fn from(salt: Message) -> [u8; Message::LEN] {
+        salt.0.into()
+    }
+}
+
+impl fmt::LowerHex for Message {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl fmt::UpperHex for Message {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl fmt::Debug for Message {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl fmt::Display for Message {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+#[cfg(feature = "std")]
+mod use_std {
+    use crate::{Hasher, Message};
+
+    use secp256k1::Message as Secp256k1Message;
+
+    impl Message {
+        pub(crate) fn to_secp(&self) -> Secp256k1Message {
+            // The only validation performed by `Message::from_slice` is to check if it is
+            // 32 bytes. This validation exists to prevent users from signing
+            // non-hashed messages, which is a severe violation of the protocol
+            // security.
+            debug_assert_eq!(Hasher::OUTPUT_LEN, secp256k1::constants::MESSAGE_SIZE);
+            Secp256k1Message::from_slice(self.as_ref()).expect("Unreachable error")
+        }
+    }
```

Comment:
A safer / no-copy conversion is available through the `ThirtyTwoByteHash` trait. This would also address the clippy convention lint that is being ignored. 

```suggestion
    impl Message {
        pub(crate) fn to_secp(self) -> Secp256k1Message {
            // The only validation performed by `Message::from_slice` is to check if it is
            // 32 bytes. This validation exists to prevent users from signing
            // non-hashed messages, which is a severe violation of the protocol
            // security.
            debug_assert_eq!(Hasher::OUTPUT_LEN, secp256k1::constants::MESSAGE_SIZE);
            Secp256k1Message::from(self)
        }
    }

    impl ThirtyTwoByteHash for Message {
        fn into_32(self) -> [u8; 32] {
            self.0.into()
        }
    }
```

Reply:
We fall into the same problem of https://github.com/FuelLabs/fuel-crypto/pull/2#discussion_r802492416 by exposing the traits of `secp256k1` publicly.

For both cases, we have the same security, but if we keep the previous code, we also keep `secp256k1` fully encapsulated.

Reply:
I see, it's probably not worth trying to optimize this by creating a private type that implements that trait until we have benchmarks and settle on whether we want to keep using this lib or k256 instead etc..